batch2/create-batch-tables.sql

```diff
@@ -94,6 +114,42 @@ CREATE INDEX batch_attributes_key_value ON `batch_attributes` (`key`, `value`(25
 
 DELIMITER $$
 
+CREATE TRIGGER attempts_before_update BEFORE UPDATE ON attempts
+FOR EACH ROW
+BEGIN
+  IF OLD.start_time IS NOT NULL AND (NEW.start_time IS NULL OR NEW.start_time > OLD.start_time) THEN
+    SET NEW.start_time = OLD.start_time;
+  END IF;
+
+  IF OLD.end_time IS NOT NULL AND (NEW.end_time IS NULL OR NEW.end_time > OLD.end_time) THEN
+    SET NEW.end_time = OLD.end_time;
+  END IF;
```

Comment:
My code handles the scenario you state above. I always pick the time point that is the earliest for both start time and end time. So in your example, we start a job running, so start is not Null, but end is Null. Then the job is preempted so we mark the end in the database and add the cost to batches and jobs. Then the end message arrives which I think must be before the preemption time (not accounting for time synchronization errors). The before update trigger says to keep the earlier time point, which will update the New.end_time to the earlier time point. The cost needs to be adjusted for the earlier time point, which I believe is what the second trigger does where it subtracts the old time interval from the new one. So in this case, the cost should be negative in the second update to jobs and batches. 

So, my reasoning is it's easier to just take the earliest time point and then you have (1) always taking precedence and then the earliest of (2) and (3) without explicitly having to do the bookkeeping. I can track the reason for the end time and that might make the code easier to always ensure (1) takes precedence, but I argue that it's not necessary. But I agree the reason would be good to have. I'll have to modify the triggers a bit to make sure I get the right reason on update.

Does this clarify my approach?

Reply:
> start is only set by the set_timing call from the worker when the attempt actually starts,

I set start from mark_complete as well. It should be the exact same time as the set_timing call. I could have logic to only set it if it's null, but you get the same answer either way.

Reply:
Just for clarity:
> which will update the New.end_time to the earlier time point

This will keep New.end_time to be New.end_time (no change in the update).